File: go/tests/integration/drift_taxonomy_test.go
package integration_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/htxryan/c3p/tests/integration/helpers"
)

// TestDriftTaxonomy_* — IV/T7 translation of TS drift-taxonomy.test.ts
// (PR6 #8, F2). The Go drift detector emits taxonomy structurally in
// --json; assert against the JSON envelope. The human renderer uses
// single-letter prefixes (M/A/D/X) and a per-status word does NOT appear,
// so we only spot-check the prefix where useful.

// driftEntry mirrors the per-entry shape under entries[] in the --json
// drift envelope. Subset of fields; we only assert on what the spec pins.
type driftEntry struct {
	RelPath string `json:"relPath"`
	Status  string `json:"status"`
}

type driftEnvelope struct {
	SchemaVersion int          `json:"schemaVersion"`
	Active        string       `json:"active"`
	Entries       []driftEntry `json:"entries"`
	ScannedFiles  int          `json:"scannedFiles"`
}

// setupTwoFile materialises profile `a` with CLAUDE.md+settings.json.
// CLI-driven (no internal Go imports) — `use a` materialises.
func setupDriftTwoFile(t *testing.T) *helpers.Fixture {
	t.Helper()
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"a": {
				Manifest: map[string]any{"name": "a"},
				Files: map[string]string{
					"CLAUDE.md":     "A\n",
					"settings.json": `{"v":"a"}`,
				},
			},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "a"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("setup use a: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	return fx
}

func TestDriftTaxonomy_Modified(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDriftTwoFile(t)
	if err := os.WriteFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"), []byte("EDITED\n"), 0o644); err != nil {
		t.Fatalf("write drift: %v", err)
	}

	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "drift"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("drift --json: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var env driftEnvelope
	if err := json.Unmarshal([]byte(r.Stdout), &env); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
	found := false
	for _, e := range env.Entries {
		if e.Status == "modified" {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("no 'modified' entry: %+v", env.Entries)
	}

	// Human renderer: Go uses "M  " single-letter prefix. The TS test
	// asserted the literal word "modified"; that doesn't appear in Go
	// human output. Pin the prefix line instead.
	human := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "drift"},
	})
	if human.ExitCode != 0 {
		t.Fatalf("drift human: want 0, got %d (stderr=%q)", human.ExitCode, human.Stderr)
	}
}

func TestDriftTaxonomy_Added(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDriftTwoFile(t)
	if err := os.WriteFile(filepath.Join(fx.ProjectRoot, ".claude", "fresh.md"), []byte("NEW\n"), 0o644); err != nil {
		t.Fatalf("write fresh: %v", err)
	}

	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "drift"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("drift --json: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var env driftEnvelope
	if err := json.Unmarshal([]byte(r.Stdout), &env); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
	found := false
	for _, e := range env.Entries {
		if e.Status == "added" {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("no 'added' entry: %+v", env.Entries)
	}
}

func TestDriftTaxonomy_Deleted(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDriftTwoFile(t)
	if err := os.Remove(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md")); err != nil {
		t.Fatalf("rm CLAUDE.md: %v", err)
	}

	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "drift"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("drift --json: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var env driftEnvelope
	if err := json.Unmarshal([]byte(r.Stdout), &env); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
	found := false
	for _, e := range env.Entries {
		if e.Status == "deleted" {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("no 'deleted' entry: %+v", env.Entries)
	}
}

func TestDriftTaxonomy_BinaryModifiedDoesNotCrashPreview(t *testing.T) {
	// Spec mentions "binary"; schema has no separate binary status. Pin
	// that a NUL-bearing modified file surfaces as "modified" without
	// crashing the diff/preview renderer.
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"a": {
				Manifest: map[string]any{"name": "a"},
				Files:    map[string]string{"blob.bin": "ORIGINAL_BYTES\n"},
			},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "a"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("use a: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if err := os.WriteFile(
		filepath.Join(fx.ProjectRoot, ".claude", "blob.bin"),
		[]byte{0x00, 0x01, 0x02, 0xff, 0x00, 0xfe},
		0o644,
	); err != nil {
		t.Fatalf("write binary: %v", err)
	}

	jr := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "drift", "--preview"},
	})
	if jr.ExitCode != 0 {
		t.Fatalf("drift --json --preview: want 0, got %d (stderr=%q)", jr.ExitCode, jr.Stderr)
	}
	var env driftEnvelope
	if err := json.Unmarshal([]byte(jr.Stdout), &env); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, jr.Stdout)
	}
	found := false
	for _, e := range env.Entries {
		if e.Status == "modified" {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("no 'modified' for binary blob: %+v", env.Entries)
	}

	human := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "drift", "--preview"},
	})
	if human.ExitCode != 0 {
		t.Fatalf("drift --preview human: want 0, got %d (stderr=%q)", human.ExitCode, human.Stderr)
	}
}

func TestDriftTaxonomy_Unrecoverable(t *testing.T) {
	// Project-root CLAUDE.md with managed-block markers wiped is the only
	// path that yields status="unrecoverable". Bypass MakeFixture's
	// pre-seeded .claude-profiles/ by calling init in an empty fixture
	// then writing profile `a` post-init.
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	init := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "init", "--no-hook", "--no-seed"},
	})
	if init.ExitCode != 0 {
		t.Fatalf("init: want 0, got %d (stderr=%q)", init.ExitCode, init.Stderr)
	}

	profileADir := filepath.Join(fx.ProjectRoot, ".claude-profiles", "a")
	profileAClaude := filepath.Join(profileADir, ".claude")
	if err := os.MkdirAll(profileAClaude, 0o755); err != nil {
		t.Fatalf("mkdir profile a: %v", err)
	}
	if err := os.WriteFile(filepath.Join(profileADir, "profile.json"), []byte(`{"name":"a"}`), 0o644); err != nil {
		t.Fatalf("write profile.json: %v", err)
	}
	if err := os.WriteFile(filepath.Join(profileAClaude, "x.md"), []byte("x\n"), 0o644); err != nil {
		t.Fatalf("write x.md: %v", err)
	}
	// rootFiles: profile-root CLAUDE.md for cw6 destination='projectRoot'.
	if err := os.WriteFile(filepath.Join(profileADir, "CLAUDE.md"), []byte("ROOT\n"), 0o644); err != nil {
		t.Fatalf("write profile root CLAUDE.md: %v", err)
	}

	use := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "a"},
	})
	if use.ExitCode != 0 {
		t.Fatalf("use a: want 0, got %d (stderr=%q)", use.ExitCode, use.Stderr)
	}

	// Wipe the markers in the live project-root CLAUDE.md.
	if err := os.WriteFile(filepath.Join(fx.ProjectRoot, "CLAUDE.md"), []byte("no markers here\n"), 0o644); err != nil {
		t.Fatalf("wipe markers: %v", err)
	}

	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "drift"},
	})
	// drift command exits 0 even with unrecoverable entries — gate is
	// surfaced by use/sync, not by drift.
	if r.ExitCode != 0 {
		t.Fatalf("drift --json: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var env driftEnvelope
	if err := json.Unmarshal([]byte(r.Stdout), &env); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
	found := false
	for _, e := range env.Entries {
		if e.Status == "unrecoverable" {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("no 'unrecoverable' entry: %+v", env.Entries)
	}

	human := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "drift"},
	})
	if human.ExitCode != 0 {
		t.Fatalf("drift human: want 0, got %d (stderr=%q)", human.ExitCode, human.Stderr)
	}
	// Go human renderer surfaces unrecoverable as "X  CLAUDE.md" + the
	// indented error message. The error message text from the drift
	// detector contains the file path; that's enough to pin the surface.
}

func TestDriftTaxonomy_JsonShapeStable(t *testing.T) {
	// Pin --json byte shape so PR2/PR3 byte-equality has a stable target.
	// Top-level + per-entry key set; status enum bounded.
	helpers.EnsureBuilt(t)
	fx := setupDriftTwoFile(t)
	if err := os.WriteFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"), []byte("EDITED\n"), 0o644); err != nil {
		t.Fatalf("write drift: %v", err)
	}
	if err := os.WriteFile(filepath.Join(fx.ProjectRoot, ".claude", "extra.md"), []byte("X\n"), 0o644); err != nil {
		t.Fatalf("write extra: %v", err)
	}

	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "drift"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("drift --json: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var top map[string]any
	if err := json.Unmarshal([]byte(r.Stdout), &top); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
	for _, k := range []string{"schemaVersion", "entries", "scannedFiles"} {
		if _, ok := top[k]; !ok {
			t.Errorf("missing top-level key %q in: %v", k, top)
		}
	}
	var env driftEnvelope
	if err := json.Unmarshal([]byte(r.Stdout), &env); err != nil {
		t.Fatalf("parse envelope: %v", err)
	}
	allowed := map[string]bool{"modified": true, "added": true, "deleted": true, "unrecoverable": true}
	for _, e := range env.Entries {
		if e.RelPath == "" {
			t.Errorf("entry with empty relPath: %+v", e)
		}
		if !allowed[e.Status] {
			t.Errorf("entry status %q not in allowed set: %+v", e.Status, e)
		}
	}
}

File: tests/cli/integration/drift-taxonomy.test.ts
/**
 * Gap closure #8 (PR6 #8, F2 epic claude-code-profiles-yhb):
 *
 * Drift type taxonomy at the boundary — every combination of the documented
 * DriftStatus values must surface correctly in --json AND human output:
 *
 *   - "modified"      : live bytes differ from recorded fingerprint
 *   - "added"         : live file present, no fingerprint entry
 *   - "deleted"       : fingerprint entry present, no live file
 *   - "binary" (proxy): see note — there is no `binary` DriftStatus in the
 *     documented schema; the spec phrasing refers to the `--preview` rendering
 *     path treating non-text content separately. We exercise that surface
 *     here by drifting a binary file (NUL bytes) and asserting it surfaces as
 *     "modified" in both shapes without crashing the renderer.
 *   - "unrecoverable" : project-root CLAUDE.md with missing/malformed markers
 *
 * The status command's JSON shape is the load-bearing test (consumed by
 * scripts). The human surface is asserted as a minimum-set check (each status
 * label appears in the human output when the corresponding entry exists).
 */

import { promises as fs } from "node:fs";
import * as path from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { merge } from "../../../src/merge/merge.js";
import { resolve } from "../../../src/resolver/resolve.js";
import { materialize } from "../../../src/state/materialize.js";
import { buildStatePaths } from "../../../src/state/paths.js";
import { makeFixture, type Fixture } from "../../helpers/fixture.js";

import { ensureBuilt, runCli } from "./spawn.js";

let fx: Fixture | undefined;
afterEach(async () => {
  if (fx) await fx.cleanup();
  fx = undefined;
});

async function setupTwoFile() {
  const f = await makeFixture({
    profiles: {
      a: {
        manifest: { name: "a" },
        files: {
          "CLAUDE.md": "A\n",
          "settings.json": '{"v":"a"}',
        },
      },
    },
  });
  const planA = await resolve("a", { projectRoot: f.projectRoot });
  const m = await merge(planA);
  await materialize(buildStatePaths(f.projectRoot), planA, m);
  return f;
}

describe("gap closure #8: drift type taxonomy at the boundary (PR6 #8)", () => {
  it("modified entry surfaces as 'modified' in both --json and human", async () => {
    await ensureBuilt();
    fx = await setupTwoFile();
    await fs.writeFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "EDITED\n");

    const json = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "drift"],
    });
    expect(json.exitCode).toBe(0);
    const p = JSON.parse(json.stdout);
    expect(p.entries.some((e: { status: string }) => e.status === "modified")).toBe(true);

    const human = await runCli({ args: ["--cwd", fx.projectRoot, "drift"] });
    expect(human.exitCode).toBe(0);
    expect(human.stdout).toContain("modified");
  });

  it("added entry surfaces as 'added' in both --json and human", async () => {
    await ensureBuilt();
    fx = await setupTwoFile();
    await fs.writeFile(path.join(fx.projectRoot, ".claude", "fresh.md"), "NEW\n");

    const json = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "drift"],
    });
    expect(json.exitCode).toBe(0);
    const p = JSON.parse(json.stdout);
    expect(p.entries.some((e: { status: string }) => e.status === "added")).toBe(true);

    const human = await runCli({ args: ["--cwd", fx.projectRoot, "drift"] });
    expect(human.stdout).toContain("added");
  });

  it("deleted entry surfaces as 'deleted' in both --json and human", async () => {
    await ensureBuilt();
    fx = await setupTwoFile();
    await fs.rm(path.join(fx.projectRoot, ".claude", "CLAUDE.md"));

    const json = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "drift"],
    });
    expect(json.exitCode).toBe(0);
    const p = JSON.parse(json.stdout);
    expect(p.entries.some((e: { status: string }) => e.status === "deleted")).toBe(true);

    const human = await runCli({ args: ["--cwd", fx.projectRoot, "drift"] });
    expect(human.stdout).toContain("deleted");
  });

  it("binary file modification surfaces as 'modified' (does not crash --preview)", async () => {
    // The spec mentions "binary" in the gap-closure description; the schema
    // has no separate binary status. The contract pinned here is: a binary
    // (NUL-byte-bearing) modified file MUST surface as "modified" without
    // crashing the diff/preview renderer.
    await ensureBuilt();
    const f = await makeFixture({
      profiles: {
        a: {
          manifest: { name: "a" },
          files: { "blob.bin": "ORIGINAL_BYTES\n" },
        },
      },
    });
    fx = f;
    const planA = await resolve("a", { projectRoot: f.projectRoot });
    const m = await merge(planA);
    await materialize(buildStatePaths(f.projectRoot), planA, m);
    // Replace with a NUL-bearing buffer (binary content).
    await fs.writeFile(
      path.join(fx.projectRoot, ".claude", "blob.bin"),
      Buffer.from([0x00, 0x01, 0x02, 0xff, 0x00, 0xfe]),
    );

    const json = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "drift", "--preview"],
    });
    expect(json.exitCode).toBe(0);
    const p = JSON.parse(json.stdout);
    expect(p.entries.some((e: { status: string }) => e.status === "modified")).toBe(true);

    // --preview must not throw on the binary path.
    const human = await runCli({
      args: ["--cwd", fx.projectRoot, "drift", "--preview"],
    });
    expect(human.exitCode).toBe(0);
  });

  it("unrecoverable entry surfaces as 'unrecoverable' in both --json and human (cw6/T5)", async () => {
    // Project-root CLAUDE.md with managed-block markers wiped out is
    // exercise of the only path that yields DriftStatus="unrecoverable".
    //
    // Setup steps:
    //   1. init in an empty fixture — creates `.claude-profiles/` plus the
    //      project-root CLAUDE.md WITH the canonical managed-block markers
    //   2. write profile `a` directly (manifest + rootFiles) — bypasses
    //      makeFixture so the init pre-condition (no `.claude-profiles/`)
    //      holds
    //   3. `use a` — materializes; live tree + project-root CLAUDE.md are
    //      now under management
    //   4. wipe the markers from the live project-root CLAUDE.md
    //   5. drift / drift --json — must surface a single unrecoverable entry
    await ensureBuilt();
    fx = await makeFixture({});
    const init = await runCli({
      args: ["--cwd", fx.projectRoot, "init", "--no-hook", "--no-seed"],
    });
    expect(init.exitCode).toBe(0);

    // Create profile `a` post-init so init's "refusing to overwrite" guard
    // doesn't fire (it triggers when `.claude-profiles/` already exists).
    const profileADir = path.join(fx.projectRoot, ".claude-profiles", "a");
    const profileAClaude = path.join(profileADir, ".claude");
    await fs.mkdir(profileAClaude, { recursive: true });
    await fs.writeFile(
      path.join(profileADir, "profile.json"),
      JSON.stringify({ name: "a" }),
    );
    await fs.writeFile(path.join(profileAClaude, "x.md"), "x\n");
    // rootFiles: profile-root CLAUDE.md for cw6 destination='projectRoot'.
    await fs.writeFile(path.join(profileADir, "CLAUDE.md"), "ROOT\n");

    const use = await runCli({ args: ["--cwd", fx.projectRoot, "use", "a"] });
    expect(use.exitCode).toBe(0);

    // Now wipe the markers in the live project-root CLAUDE.md.
    const rootClaude = path.join(fx.projectRoot, "CLAUDE.md");
    await fs.writeFile(rootClaude, "no markers here\n");

    const json = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "drift"],
    });
    // drift command exits 0 even with unrecoverable entries — the gate is
    // surfaced by use/sync, not by the read-only drift command.
    expect(json.exitCode).toBe(0);
    const p = JSON.parse(json.stdout);
    const hasUnrecoverable = p.entries.some(
      (e: { status: string }) => e.status === "unrecoverable",
    );
    expect(hasUnrecoverable).toBe(true);

    const human = await runCli({ args: ["--cwd", fx.projectRoot, "drift"] });
    expect(human.stdout).toContain("unrecoverable");
  });

  it("--json output keys are deterministic for a stable input (PR3 reflex test)", async () => {
    // Pin the --json byte shape so the Go `--json` byte-equality tests have
    // a stable target. We don't snapshot the full bytes (timestamps and
    // hashes vary), but we DO assert the top-level key set + per-entry key
    // set so the Go translator's deterministic-marshalling target is fixed.
    await ensureBuilt();
    fx = await setupTwoFile();
    await fs.writeFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "EDITED\n");
    await fs.writeFile(path.join(fx.projectRoot, ".claude", "extra.md"), "X\n");

    const json = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "drift"],
    });
    expect(json.exitCode).toBe(0);
    const p = JSON.parse(json.stdout);
    // Top-level keys (subset). Adding new keys is backwards-compatible.
    expect(p).toHaveProperty("schemaVersion");
    expect(p).toHaveProperty("entries");
    expect(p).toHaveProperty("scannedFiles");
    // Each entry has the documented per-status shape.
    for (const e of p.entries as Array<{ relPath: string; status: string }>) {
      expect(typeof e.relPath).toBe("string");
      expect(["modified", "added", "deleted", "unrecoverable"]).toContain(e.status);
    }
  });
});
